packages/viewer/src/scenes/operations.ts

```diff
@@ -0,0 +1,241 @@
+
+  public withItemId(partId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withItemId(partId));
+    return this;
+  }
+}
+
+/**
+ * An `ItemSelector` describes a type for performing a selection against some
+ * attribute of a scene item.
+ */
+export type ItemSelector = SuppliedIdSelector | ItemIdSelector;
+
+interface OrSelector {
+  type: 'or';
+  selectors: ItemSelector[];
+}
+
+interface AndSelector {
+  type: 'and';
+  selectors: ItemSelector[];
+}
+
+export interface OperationDefinition {
+  operation: ItemOperation | SceneOperation;
+}
+
+export interface SelectorBuilder<T> {
+  /**
+   * Returns the built selector.
+   */
+  build(): T;
+
+  /**
+   * Returns a conditional builder to perform an `or` operation.
+   */
+  or(): SelectorBuilder<OrSelector>;
+
+  /**
+   * Returns a conditional builder to perform an `or` operation.
+   */
+  and(): SelectorBuilder<AndSelector>;
+
+  /**
+   * Selects a item that has the given Customer facing ID
+   *
+   * @param suppliedId The customer ID of an item to match
+   */
+  withSuppliedId(suppliedId: string): this;
+
+  /**
+   * Selects a item that has the given Vertex ID.
+   *
+   * @param itemId The ID of the item to match.
+   */
+  withItemId(itemId: string): this;
+}
+
+interface ChangeMaterialOperation {
+  type: 'change-material';
+  color: string;
+}
+
+export type SceneOperation = ChangeMaterialOperation;
+
+export type ItemOperation = ShowItemOperation | HideItemOperation;
+
+export interface SceneItemQuery {
+  where(
+    query: SelectorBuilder<ItemSelectorBuilder>
+  ): SceneItemOperationsExecutor;
+}
+
+export interface SceneItemOperations<T> {
+  material(color: string): T;
+  show(): T;
+  hide(): T;
+}
+
+export class SceneOperationBuilder
+  implements SceneItemOperations<SceneOperationBuilder> {
+  private operations: OperationDefinition[] = [];
+
+  /**
+   * Constructs the scene operations and returns a definition describing each
+   * operation.
+   */
+  public build(): OperationDefinition[] {
+    return this.operations.concat();
+  }
+
+  public material(color: string): SceneOperationBuilder {
+    return this.operation({ type: 'change-material', color });
+  }
+
+  public show(): SceneOperationBuilder {
+    return this.operation({ type: 'show' });
+  }
+
+  public hide(): SceneOperationBuilder {
+    return this.operation({ type: 'hide' });
+  }
+
+  private operation(operation: SceneOperation): SceneOperationBuilder;
+  private operation(operation: ItemOperation): SceneOperationBuilder;
+
+  private operation(...args: any[]): this {
+    const operation = args[0];
+    if (args.length === 1) {
+      this.operations.push({ operation });
+    } else if (args.length === 2) {
+      this.operations.push({ operation });
+    }
+    return this;
+  }
+}
+
+export class SceneQueryBuilder implements SceneItemQuery {
+  private query: SelectorBuilder<ItemSelectorBuilder> | undefined;
+
+  public where(
+    _query: SelectorBuilder<ItemSelectorBuilder>
```

Comment:
if there's a naming collision, convention is to make the member variable use a `_`.